### PR TITLE
sql/catalog: introduce external catalog pkg

### PR DESCRIPTION
**Pull request by msbutler**
This patch adds a bare bones api which LDR will use to extract table
descriptors from the source cluster and ingest them into the destination
cluster before beginning a fast initial scan.

Epic: none

Release note: none

### CODE REVIEW

**Comment by msbutler** [comment]
unrelated flake in unit test job

File: pkg/crosscluster/externalcatalog/external_catalog.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+	"github.com/cockroachdb/cockroach/pkg/sql/parser"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/errors"
+)
+
+// ExtractExternalCatalog extracts the table descriptors via the schema
+// resolver.
+func ExtractExternalCatalog(
+	ctx context.Context, schemaResolver resolver.SchemaResolver, tableNames ...string,
+) (*externalpb.ExternalCatalog, error) {
```
Comment (dt) [inline pkg/crosscluster/externalcatalog/external_catalog.go:30]: doesn't need to be a pointer type does it?

File: pkg/crosscluster/externalcatalog/external_catalog.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+// IngestExternalCatalog ingests the tables in the external catalog into into
+// the database and schema.
+//
+// TODO: provide a list of databaseID/schemaID pairs to ingest into.
+func IngestExternalCatalog(
+	ctx context.Context,
+	execCfg *sql.ExecutorConfig,
+	user username.SQLUsername,
+	externalCatalog *externalpb.ExternalCatalog,
```
Comment (dt) [inline pkg/crosscluster/externalcatalog/external_catalog.go:57]: no reason to be a pointer here right? caller doesn't modify the content of the passed value

File: pkg/crosscluster/externalcatalog/external_catalog.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+	databaseID descpb.ID,
+	schemaID descpb.ID,
+	setOffline bool,
+) error {
+	// TODO: consider pulling the DescTxn outside of this function.
```
Comment (dt) [inline pkg/crosscluster/externalcatalog/external_catalog.go:62]: yeah, I'd pass this a isql.Txn and descs.Collection instead

File: pkg/crosscluster/externalcatalog/external_catalog.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+	return sql.DescsTxn(ctx, execCfg, func(
+		ctx context.Context, txn isql.Txn, descsCol *descs.Collection,
+	) error {
+		dbDesc, err := descsCol.ByIDWithLeased(txn.KV()).WithoutNonPublic().Get().Database(ctx, databaseID)
+		if err != nil {
+			return err
+		}
+		tablesToWrite := make([]catalog.TableDescriptor, 0)
```
Comment (dt) [inline pkg/crosscluster/externalcatalog/external_catalog.go:70]: len(externalCatalog.Tables)

File: pkg/crosscluster/externalcatalog/external_catalog.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+		externalCatalog.Tables = append(externalCatalog.Tables, td.TableDescriptor)
+	}
+	return &externalCatalog, nil
+}
+
+// IngestExternalCatalog ingests the tables in the external catalog into into
+// the database and schema.
+//
+// TODO: provide a list of databaseID/schemaID pairs to ingest into.
+func IngestExternalCatalog(
```
Comment (dt) [inline pkg/crosscluster/externalcatalog/external_catalog.go:53]: This feels ingestion behavior, not external behavior, so maybe this function should be in `sql/catalog/ingesting` package instead?

**Review by fqazi — COMMENTED** [review]
Overall this looks pretty good, just a comment and question.

> Reviewed 4 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136770#-:-ODNX-xZ1HCxG9rwWy9O:b-8io7u8)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jeffswenson and @msbutler)

___
*[`pkg/crosscluster/externalcatalog/external_catalog.go` line 81 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136770#-ODNUt5KE_8P0i5WLnq3:-ODNUt5KE_8P0i5WLnq4:b-n5yvbc) ([raw file](https://github.com/cockroachdb/cockroach/blob/146b096d812541da7cc81f59c7d55c0f2fe24f58/pkg/crosscluster/externalcatalog/external_catalog.go#L81)):*
> ```Go
> 			mutTable := tabledesc.NewBuilder(&table).BuildCreatedMutableTable()
> 			if setOffline {
> 				mutTable.SetOffline("")
> ```

Probably as TODO we want some way to set the offline reason
___
*[`pkg/crosscluster/externalcatalog/external_catalog_test.go` line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136770#-ODNWXIt0DpIMKblzhzw:-ODNWXIu78nEM6ZgQcMQ:b1cb0e4) ([raw file](https://github.com/cockroachdb/cockroach/blob/146b096d812541da7cc81f59c7d55c0f2fe24f58/pkg/crosscluster/externalcatalog/external_catalog_test.go#L93)):*
> ```Go
> 	}))
> 
> 	require.ErrorContains(t, IngestExternalCatalog(ctx, &execCfg2, sqlUser, sadCatalog, parentID, schemaID, false), "invalid foreign key backreference")
> ```

Do we want to even allow partial extractions? We could assert this during the extraction process via GetAllReferenced.* functions to detect problems.

File: pkg/crosscluster/externalcatalog/external_catalog.go (hunk 4 above)
Comment (msbutler) [reply pkg/crosscluster/externalcatalog/external_catalog.go]: Done.

**Review by msbutler — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136770)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @fqazi, and @jeffswenson)

___
*[`pkg/crosscluster/externalcatalog/external_catalog.go` line 81 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136770#-ODNUt5KE_8P0i5WLnq3:-ODNxuL8Dxrwb8bR-5oJ:b-v7gbdv) ([raw file](https://github.com/cockroachdb/cockroach/blob/146b096d812541da7cc81f59c7d55c0f2fe24f58/pkg/crosscluster/externalcatalog/external_catalog.go#L81)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Probably as TODO we want some way to set the offline reason
</blockquote></details>

yeah, i was envisioning creating some functional option args once we actually started using it.
___
*[`pkg/crosscluster/externalcatalog/external_catalog_test.go` line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136770#-ODNWXIt0DpIMKblzhzw:-ODNy4XP4ikkVBluP3Jl:b-tnf9pb) ([raw file](https://github.com/cockroachdb/cockroach/blob/146b096d812541da7cc81f59c7d55c0f2fe24f58/pkg/crosscluster/externalcatalog/external_catalog_test.go#L93)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Do we want to even allow partial extractions? We could assert this during the extraction process via GetAllReferenced.* functions to detect problems.
</blockquote></details>

yes, we allow partial extractions. Is Validate called at the end of the DescTxn, to ensure that this func will never break the schema graph?

**Review by fqazi — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136770)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @jeffswenson, and @msbutler)

___
*[`pkg/crosscluster/externalcatalog/external_catalog.go` line 81 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136770#-ODNUt5KE_8P0i5WLnq3:-ODO15kSDjCeZw5j8LB-:b-hfghm4) ([raw file](https://github.com/cockroachdb/cockroach/blob/146b096d812541da7cc81f59c7d55c0f2fe24f58/pkg/crosscluster/externalcatalog/external_catalog.go#L81)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

yeah, i was envisioning creating some functional option args once we actually started using it.
</blockquote></details>

Sounds good
___
*[`pkg/crosscluster/externalcatalog/external_catalog_test.go` line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136770#-ODNWXIt0DpIMKblzhzw:-ODO12U82pDDKj0u583n:b-p5vnto) ([raw file](https://github.com/cockroachdb/cockroach/blob/146b096d812541da7cc81f59c7d55c0f2fe24f58/pkg/crosscluster/externalcatalog/external_catalog_test.go#L93)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

yes, we allow partial extractions. Is Validate called at the end of the DescTxn, to ensure that this func will never break the schema graph?
</blockquote></details>

Yeah, that safety is still therr

File: pkg/crosscluster/externalcatalog/external_catalog.go (hunk 5 above)
Comment (msbutler) [reply pkg/crosscluster/externalcatalog/external_catalog.go]: per discussion offline, i moved everything the to `sql/catalog`. Note that moving externalpb over to catpb does not work as descpb imports catpb.

**Comment by msbutler** [comment]
TFTR!

bors r=fqazi

**Pull request merged**